operator/v1/types_klusterlet.go

```diff
@@ -87,25 +90,31 @@ type KlusterletSpec struct {
 	// +optional
 	HubApiServerHostAlias *HubApiServerHostAlias `json:"hubApiServerHostAlias,omitempty"`
 
-	// ResourceRequirement specify QoS classes of klusterlet deployment
+	// ResourceRequirement specify QoS classes of deployments managed by klusterlet.
+	// It applies to all the containers in the deployments.
 	// +optional
 	ResourceRequirement *ResourceRequirement `json:"resourceRequirement,omitempty"`
 }
 
 type ResourceQosClass string
 
 const (
-	// Default use resource setting in the template file
+	// Default use resource setting in the template file (with requests but no limits in the resources)
 	ResourceQosClassDefault ResourceQosClass = "Default"
 	// If all containers in the pod don't set resource request and limits, the pod is treated as BestEffort.
 	ResourceQosClassBestEffort ResourceQosClass = "BestEffort"
+	// Configurable resource requirements with requests and limits
+	ResourceQosClassResourceRequirement ResourceQosClass = "ResourceRequirement"
 )
 
 // ResourceRequirement allow user override the default pod QoS classes
 type ResourceRequirement struct {
-	// +kubebuilder:validation:Enum=Default;BestEffort
+	// +kubebuilder:validation:Enum=Default;BestEffort;ResourceRequirement
 	// +kubebuilder:default:=Default
 	Type ResourceQosClass `json:"type"`
+	// Resources defines resource requests and limits when Type is ResourceQosClassResourceRequirement
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
```

Comment:
```suggestion
	ResourceRequirement *corev1.ResourceRequirements `json:"resourceRequirement,omitempty"`
```

Reply:
if this change is made, the yaml will look like:
```
resourceRequirement:
  resourceRequirement:
    limits: 
       ...
```
duplicated `resourceRequirement`s look weild I suppose

Reply:
In kubernetes Container struct, it is
```
Resources ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,8,opt,name=resources"`
```

Reply:
it is intended to follow the union pattern https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md

Reply:
updated